cmd/pineconesim/simulator/router.go

```diff
@@ -68,3 +67,176 @@ func (r *DefaultRouter) ConfigureFilterPeer(peer types.PublicKey, rates adversar
+		destination: a,
+		hops:        1,
+	}
+
+	nexthop := r.rtr.NextHop(nil, frameType, a)
+	if nexthop == nil {
+		return 0, 0, fmt.Errorf("no valid nexthop for ping")
+	}
+
+	p := make([]byte, 256)
+	_, err := payload.MarshalBinary(p)
+	if err != nil {
+		return 0, 0, fmt.Errorf("failed marshalling ping payload: %w", err)
+	}
+
+	_, writeErr := r.rtr.WriteTo(p, *nexthop)
+	if writeErr != nil {
+		return 0, 0, fmt.Errorf("failed sending ping to node: %w", writeErr)
+	}
+
+	start := time.Now()
+	v, existing := r.pings.LoadOrStore(id, make(chan uint16))
+	if existing {
+		return 0, 0, fmt.Errorf("a ping to this node is already in progress")
+	}
+	defer r.pings.Delete(id)
+	ch := v.(chan uint16)
+	select {
+	case <-ctx.Done():
+		return 0, 0, fmt.Errorf("ping timed out")
+	case hops := <-ch:
+		return hops, time.Since(start), nil
+	}
+}
+
+func (r *DefaultRouter) OverlayReadHandler(quit <-chan bool) {
+	buf := make([]byte, 256)
+	for {
+		select {
+		case <-quit:
+			return
+		default:
+		}
+
+		_, addr, err := r.rtr.ReadFrom(buf)
+		if err != nil {
+			continue
+		}
+
+		payload := PingPayload{}
+		_, pingErr := payload.UnmarshalBinary(buf)
+		if pingErr != nil {
+			println(pingErr.Error())
+			continue
+		}
+
+		pingAtDest := false
+		var frameType types.FrameType
+		switch payload.pingType {
+		case TreePing:
+			switch dest := (payload.destination).(type) {
+			case types.Coordinates:
+				frameType = types.TypeTreeRouted
+				if dest.EqualTo(r.Coords()) {
+					pingAtDest = true
+				}
+			}
+		case TreePong:
+			switch orig := (payload.origin).(type) {
+			case types.Coordinates:
+				frameType = types.TypeTreeRouted
+				if orig.EqualTo(r.Coords()) {
+					id := payload.destination.String()
+					v, ok := r.pings.Load(id)
+					if !ok {
+						continue
+					}
+					ch := v.(chan uint16)
+					ch <- payload.hops
+					close(ch)
+					r.pings.Delete(id)
+					continue
+				}
+			}
+		case SNEKPing:
+			switch dest := (payload.destination).(type) {
+			case types.PublicKey:
+				frameType = types.TypeVirtualSnakeRouted
+				if dest == r.PublicKey() {
+					pingAtDest = true
+				}
+			}
+		case SNEKPong:
+			switch orig := (payload.origin).(type) {
+			case types.PublicKey:
+				frameType = types.TypeVirtualSnakeRouted
+				if orig == r.PublicKey() {
+					id := payload.destination.String()
+					v, ok := r.pings.Load(id)
+					if !ok {
+						continue
+					}
+					ch := v.(chan uint16)
+					ch <- payload.hops
+					close(ch)
+					r.pings.Delete(id)
+					continue
+				}
+			}
+		default:
+			continue
+		}
+
+		var fromAddr *net.Addr
+		fromAddr = &addr
+		if payload.pingType == TreePing || payload.pingType == SNEKPing {
+			if !pingAtDest {
+				payload.hops++
+			} else {
+				fromAddr = nil
+				if frameType == types.TypeTreeRouted {
+					payload.pingType = TreePong
+				} else {
+					payload.pingType = SNEKPong
+				}
+			}
+		}
+
+		_, pErr := payload.MarshalBinary(buf)
+		if pErr != nil {
+			continue
+		}
+
+		var nexthop *net.Addr
```

Comment:
`net.Addr` is an interface — it already has the potential to be `nil`-able. Taking pointers to `interface`s is like taking pointers to pointers :-)